### ISSUE TITLE
[RFC] Node.js v16 Upgrade Plan

### ISSUE THREAD

**Issue by boktorbb-amzn** [opensearch-project/OpenSearch-Dashboards#835]
<h2>Overview</h2>OpenSearch Dashboards is currently running an EOL Node.js version, Node.js v10.24.1. This current version of Node carries with it several high severity security vulnerabilities, as well as forces us to use older version of critical dependencies that carry their own share security vulnerabilities. The intent of the core Dashboards team is to upgrade to Node.js v16.9.1 which goes into LTS support in October 2021. The OpenSearch project is following the SemVer standard for versioning; therefore, the Node.js v16 upgrade is considered a breaking change and will require a version bump of Dashboards to 2.0.0 <hr/><h2>Scope</h2>The scope of the Node 16 upgrade is to do the minimum amount of dependency, code, and testing changes to allow for the core Dashboards team to address many of the CVEs that are present currently and blocked by our current Node.js version. While there may be many opportunities to make changes or upgrade packages because of the new version of Node, they would be out of scope of the work done to upgrade Dashboards to Node.js v16. <br><hr/><h2>Upgrade Proposal</h2>The Node.js v16 upgrade will take the form of 3 phases that will be cycled through for each new plugin and developer wanting to upgrade. The 3 phases are somewhat coupled which makes working on the phases asynchronously difficult.<br><h3>Phase 1: Dependency Version Parity</h3>This phase consists primarily of bringing all of Dashboards’ dependencies up to parity with what Node.js v16 expects/requires. Node v16 is many major versions higher than what we were working with before and will require upgrades to several dependencies. These changes will introduce cascading breaking changes to other dependencies, as well as to some code implementations in Dashboards.<br><br>Below is a table of dependencies that are required to upgrade to Node.js v16 for core Dashboards:<br><br>

Dependencies | Current Version | Upgrade Version | Notes | Function in Dashboards
-- | -- | -- | -- | --
lmdb-store | 0.6.10 | 1.6.7 | The primary issue introduced by Node.js v16 upgrade is that the sub-dependency `msgpackr-extract` is not supported by Node 16 and requires lmdb-store to be bumped to latest. | lmdb-store is a key-value store library being used by our `osd-optimizer` package for cluster caching
node-sass | v5 | 6.0.1 | node-sass is deprecated and should be changed to dart-sass; however, there was a version of node-sass that was supported by Node.js v16 and that was version 6.0.1. Bumping node-sass rather than swapping out to dart-sass seems to be less work and there are other open questions surrounding the use of dart-sass with regards to EUI. | node-sass is the primary library for using Sass. Currently deprecated in favor of dart-sass
sass-loader | 8.0.2 | 10.2.0 | Bumping node-sass up to v6.0.1 causes the need to upgrade sass-loader to a newer version that supports both node-sass v6.0.1 and node.js v16.9.1. | helper library for all things Sass
@types/webpack-env | 1.15.2 | 1.15.3 | This version bump was required because of a bug with Node.js interfacing with  webpack. The error was: `error TS2430: Interface 'NodeJS.Module' incorrectly extends interface '__WebpackModuleApi.Module` | Type definitions for typescript that correspond to webpack and are to be used in conjunction with webpack
tslib | N/A | 2.0.0 | The tslib module is explicitly required for Node.js v16 and was not added in our package.json | runtime library for typescript that contains all of the typescript helper functions
@types/node | >=10.17.17 <10.20.0 | 16.9.1 | It's required to update to the latest type definition package for node that corresponds to the Node.js major version being used | Type definitions for typescript that correspond to Node major versions

The majority of the errors pertaining to dependency version errors and version mismatches will manifest themselves as build errors and package compatibility errors. Some of the errors will manifest as type/code logical errors but those will be handled in the next phase.<br><h3>Phase 2: Breaking code changes</h3>The next major obstacle is the plethora of breaking code changes introduced both by the new version of Node.js itself and the new versions of other dependencies. This phase requires a good amount of familiarity with breaking changes across major versions of dependencies. The three major types of code changes are listed below:<br><br><ol><li>Syntax strictness with regards to null and undefined behavior</li><ol><li>Node.js v16 has become much more strict with the handling of possible null and undefined behaviors.</li><li>The majority of these types of errors are fixed by passing the <code>non-null assertion operator</code> if you are sure that the value being passed in code is not going to be null</li><ol><li>e.g. </li><li> childProcess.stdin!.end(stdin, 'utf8'); // TypeScript note: As long as the proc stdio[1] is 'pipe', then stdin will not be null</li><li>Node has an issue with stdin and stderr null possibilities and requires the <code>non-null assertion operator</code> to no longer break the build</li></ol></ol><li>Type mismatches and type strictness</li><ol><li>There are lambdas throughout the code base that are using variables with no explicit type</li><li>Variables being typed as unknown so they break type definitions for certain interfaces</li><li>Generally the error messages take the form of <code>Type&lt;unknown&gt; cannot be assigned to Type&lt;type&gt;</code></li><li>There are cases of the type definitions expected by <code>@types/node</code> not matching the actual types</li><ol><li>E.g. </li><li>`urlUtils.encodeQuery` expects key/value pairs with values of type `string | string[] | null`,<br>however `@types/node` says that `url.query` has values of type `string | string[] | undefined`.<br>After investigating this, it seems that no matter what the values will be of type `string | string[]`</li></ol></ol><li>Feature deprecation and new data interfaces</li><ol><li>Webpack config file has deprecated an object property in favor of another in newer versions</li><ol><li>webpack-config: <code>prependData</code> changed to <code>additionalData</code></li><li>The <code>additionalData</code> property requires an additional <code>content</code> parameter</li></ol><li>The <code>lmdb-store</code> version bump leads to a required rewrite of our <code>cache.ts</code> file in the <code>osd-optimizer</code> package</li><ol><li>This is required because in v0.6.10, there was no data interface for the store. This lead to elastic creating their own data interface abstraction</li><li>interface Lmdb&lt;T&gt; {<br>  get(key: string): T | undefined;<br>  put(key: string, value: T, version?: number, ifVersion?: number): Promise&lt;boolean&gt;;<br>  remove(key: string, ifVersion?: number): Promise&lt;boolean&gt;;<br>  openDB&lt;T&gt;(options: { name: string; encoding: 'msgpack' | 'string' | 'json' | 'binary' }): Lmdb&lt;T&gt;;<br>  getRange(options?: {<br>    start?: T;<br>    end?: T;<br>    reverse?: boolean;<br>    limit?: number;<br>    versions?: boolean;<br>  }): Iterable&lt;{ key: string; value: T }&gt;;<br>}</li><li>In <code>lmdb-store</code> v1.6.7, they have introduced new interfaces that must be used in our caching implementation.</li><li>The <code>lmdb-store</code> types are:</li><ol><li><code>LmdbStore.RootDatabase</code></li><li><code>LmdbStore.Database</code></li></ol><li> A rewrite of some of the member methods of the <code>Cache</code> class is required to satisfy the more strict exception handling requirements of Node</li></ol></ol></ol><h3>Phase 3: Runtime Errors</h3>Runtime errors tend to be the hardest types of errors to track down in this process and they usually occur either as you attempt to run Dashboards using <code>yarn start</code> or they crop up while Dashboards is running itself. <br><br><i><b>Node 16 Cluster error:</b></i><br><br>There seems to be an issue with importing the <code>cluster</code> module in the <code>worker.ts</code> file in core. <br><br><pre> log   [16:54:15.757] [fatal][root] TypeError: _cluster.default.setupPrimary is not a function</pre>When switching the import from <code>import cluster from 'cluster'</code> to <code>import * as cluster from 'cluster'</code> we get the following error.<br><pre>Unhandled Promise rejection detected:<br><br>TypeError: Cannot read properties of undefined (reading 'apply')<br>    at BinderFor.on (/home/ubuntu/OpenSearch-Dashboards/src/cli/cluster/binder.ts:47:8)<br>    at BinderFor.on (/home/ubuntu/OpenSearch-Dashboards/src/cli/cluster/binder_for.ts:41:11)<br>    at Worker.start (/home/ubuntu/OpenSearch-Dashboards/src/cli/cluster/worker.ts:228:24)<br>    at ClusterManager.startCluster (/home/ubuntu/OpenSearch-Dashboards/src/cli/cluster/cluster_manager.ts:181:14)<br>    at FSWatcher.&lt;anonymous&gt; (/home/ubuntu/OpenSearch-Dashboards/src/cli/cluster/cluster_manager.ts:249:12)</pre><br><i><b>API Error</b></i>:<br><br>The following 2 errors occur when starting dashboards with <code>yarn start</code> after upgrading <code>lmdb-store</code>, <code>node-sass</code>, and <code>sass-loader</code>.  It stops Dashboards from loading the home app because the capabilities api call never succeeds.<br><pre><code>NetworkError when attempting to fetch resource.<br>Version</code><code>:</code><code> </code><code>1.1</code><code>.</code><code>0</code><br><code>Build</code><code>:</code><code> </code><code>9007199254740991</code><br><code>Error</code><code>:</code><code> </code><code>NetworkError</code><code> </code><code>when</code><code> attempting to fetch resource</code><code>.</code><br><code>HttpFetchError@http</code><code>:</code><code>//localhost:5603/ime/9007199254740991/bundles/core/core.entry.js:18398:5</code><br><code>fetchResponse@http</code><code>:</code><code>//localhost:5603/ime/9007199254740991/bundles/core/core.entry.js:18274:13</code></pre><pre>Uncaught (in promise) TypeError: NetworkError when attempting to fetch resource.<br>HttpFetchError http_fetch_error.ts:38<br> fetchResponse fetch.ts:156<br> __osdBootstrap__ osd_bootstrap.ts:57<br> AsyncFunctionThrow self-hosted:697</pre>Looks like the above errors pertain to the api/core/capabilities API request <br><pre>POST http://localhost:5603/azc/api/core/capabilities net::ERR_EMPTY_RESPONSE</pre><hr/>

**Comment by dblock** [comment]
Please make sure to include changes to https://github.com/opensearch-project/opensearch-build when node is upgraded. It likely means building different docker images.

**Comment by AMoo-Miki** [comment]
@stockholmux, I hear you; it helps that the upgrade will come some time after v16 goes to LTS. 

Node will always have 2 or 3 LTS versions simultaneously and because I would love to avoid heartburn for everyone, I would prefer to invest a bit more effort and make Dashboards run on all the LTS versions that people might be running with.

**Comment by boktorbb-amzn** [comment]
Node upgrade progress update:

Overview

OpenSearch Dashboards needs to upgrade the Node.js version from the current version, 10.24.1, to mitigate high severity CVEs affecting the application. The upgrade process intended to move to Node.js v16.9.1. During the upgrade process, it became clear that the amount of investigation and work required to upgrade to v16 was quite large and the team wanted to prioritize mitigating CVEs for the next OpenSearch release. Since the Node v16 upgrade wouldn’t be finished in time for the next release, the team decided on a move to Node.js v14 instead.

Issues faced

The jump to v16 is several major versions higher than the current Dashboards version. Particularly, the jump from v14 to v16 is a much bigger jump in terms of breaking changes than even the jump to v10 to v14. Node v16 has many breaking changes and we encountered the following issues:

* The first major issue is dependency mismatches. Dependencies have silent errors and breaking changes when we update the mandatory dependencies that are required to match Node.js v16.
    * One of the major issues of these dependency updates is the changes in types and interfaces. When they change, it requires us to refactor code and modules in Dashboards. This is mitigated by moving back to Node v14 where these breaking interface changes are minimal
    * The other major issue faced was silent failures during runtime. Certain dependencies are mismatched with others but the mismatch doesn’t manifest in errors or log information but causes silent runtime crashes and failures. This is mitigated by moving to Node v14 where the package changes and updates are minimized
* The next major issue is code refactor in the Dashboards repo
    * The cluster module in the Dashboards core would require a refactor to mitigate the changes in the Node.js v16 built-in cluster module API changes.
        * A refactor of the way we determine worker types and scheduling would be required
        * A refactor of the osd-optimizer process that parses plugin paths and loads plugins would be required
* Another issue is the extra effort that is required from individual plugin owners to upgrade their plugins for compatibility with Node v16. The jump from v10 to v16 is big enough that the amount of work required from each plugin can be large. With 9 internal Dashboards plugins being affected, moving to Node v14 would reduce effort on this front significantly.


Possible paths forward

Moving to Node 14 (preferred solution)

Node.js v14 is currently in LTS and will continue to be in LTS until 2023. This gives us an easier and earlier target to hit since we’re prioritizing mitigating CVEs. It also gives us time to plan and prioritize work for a future upgrade to Node v16 where we aren’t crunched for time because of CVEs and meeting predetermined release schedules.

Continuing Node 16 and missing OpenSearch release

Another option would be bringing on more people to help with the Node 16 upgrade. This would take more effort and would require us to miss the OpenSearch 1.3 release and wait until the following release. Since mitigating the high severity CVEs in Dashboards motivated the whole Node upgrade project, pushing off the CVE fixes further out while Node v10 is EOL is not a good option

Continuing Node 16 and delaying OpenSearch release

This option is the same as the above option but instead of missing the release date and waiting until the next one, we have the OpenSearch 1.3 release delayed until Node.js v16 upgrade is complete.

**Comment by seanneumann** [comment]
Thanks for the research and prototyping @boktorbb-amzn. I think moving to 14 sounds like a good next step.

**Comment by stockholmux** [comment]
So, 14 and 16 are breaking changes, so Dashboards will jump majors twice to get to the latest version?

What's the timeline/target distribution version for 16?

**Comment by seanneumann** [comment]
According to [Node.js LTS schedule](https://nodejs.org/en/about/releases/), 14 is EOL April 2023. My hunch is we'd upgrade in Q1 of 2023.

**Comment by stockholmux** [comment]
@seanneumann That makes sense, if we could advance it a tad, there would be a nice symmetry of V14 = 2022, V16 = 2023.

**Comment by stockholmux** [comment]
And naturally, I think everyone would like to avoid releasing Dashboards with a EOL Node ever again, so leaving to the last part of the support window is probably not preferred.

**Comment by boktorbb-amzn** [comment]
Node upgrade to version 14.18.2 is complete. Will close this issue for now and will reopen when the time comes to evaluate the next version to upgrade to for OpenSearch Dashboards
